### ISSUE TITLE
Soundbridge selective transcoding

### ISSUE THREAD

**Issue by frankusb** [owntone/owntone-server#1665]
1) Could httpd_rsp.c transcode to CBR MP3 instead of WAV? This is mostly a question of if ffmpeg can do FLAC to MP3 like it does FLAC to WAV now.
2) If so, could it do it on a per client basis?

Wireless Soundbridges can't handle WAV bitrates. Currently I have two instances of owntone running, one in conjunction with mp3fs to allow reliable playback on wireless Soundbridges. Besides the duplication of resources, the real annoyance is when I update a playlist, the mp3fs owntone then churns for 30-40 minutes re-reading all of the metadata since inotify is not supported with mp3fs.

**Comment by ejurgensen** [comment]
Sounds like it should be possible, I'll look into it.

**Comment by ejurgensen** [comment]
For testing this concept, I have made a branch with a very rudimentary implementation where the transcoding is switched to CBR MP3 at 256 kbit/s. Could you try if this works with a flac and the Soundbridge?

This is the branch: https://github.com/owntone/owntone-server/tree/rspmp3

**Comment by frankusb** [comment]
I get httpd: Transcoding error, file id 37484 in the log.

**Comment by ejurgensen** [comment]
Which version of ffmpeg is it?

Also, can you confirm it doesn't play and you get the same error if you try from a browser? Try with http://[address]:3689/rsp/stream/37484

**Comment by frankusb** [comment]
ffmpeg version 6.0 Copyright (c) 2000-2023 the FFmpeg developers
  built with gcc 12 (Debian 12.2.0-14)
  configuration: --disable-decoder=amrnb --disable-decoder=libopenjpeg --disable-gnutls --disable-liblensfun --disable-libopencv --disable-podpages --disable-sndio --disable-stripping --enable-avfilter --enable-chromaprint --enable-frei0r --enable-gcrypt --enable-gpl --enable-ladspa --enable-libaom --enable-libaribb24 --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libdav1d --enable-libdavs2 --enable-libdc1394 --enable-libdrm --enable-libfdk-aac --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libglslang --enable-libgme --enable-libgsm --enable-libiec61883 --enable-libjack --enable-libkvazaar --enable-libmp3lame --enable-libmysofa --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenh264 --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librabbitmq --enable-librist --enable-librsvg --enable-librubberband --enable-libshine --enable-libsmbclient --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvo-amrwbenc --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxavs2 --enable-libxml2 --enable-libxvid --enable-libzimg --enable-libzmq --enable-libzvbi --enable-lv2 --enable-nonfree --enable-omx --enable-openal --enable-opencl --enable-opengl --enable-openssl --enable-postproc --enable-pthreads --enable-shared --enable-version3 --enable-vulkan --incdir=/usr/include/x86_64-linux-gnu --libdir=/usr/lib/x86_64-linux-gnu --prefix=/usr --toolchain=hardened --enable-vaapi --enable-libvpl --enable-libvmaf --enable-libilbc --enable-libjxl --disable-altivec --shlibdir=/usr/lib/x86_64-linux-gnu
  libavutil      58.  2.100 / 58.  2.100
  libavcodec     60.  3.100 / 60.  3.100
  libavformat    60.  3.100 / 60.  3.100
  libavdevice    60.  1.100 / 60.  1.100
  libavfilter     9.  3.100 /  9.  3.100
  libswscale      7.  1.100 /  7.  1.100
  libswresample   4. 10.100 /  4. 10.100
  libpostproc    57.  1.100 / 57.  1.100
Hyper fast Audio and Video encoder

It does not play and I get the same error if I try from a browser.

I also noticed that the Soundbridge shows the song info as "Kind:WAV audio file" and "Bit Rate: 1411 kbps" for FLAC files which is not what I would expect.

**Comment by ejurgensen** [comment]
Ok, I found the reason it's not playing in the browser. I'll get back to you with a new attempt.

**Comment by ejurgensen** [comment]
I've pushed a commit to the branch that I think should fix at least playback in browser. Please test and let me know.

**Comment by frankusb** [comment]
I won't be able to test until Wednesday.

**Comment by frankusb** [comment]
The Soundbridge is unable to stream. It still shows shows the song info as "Kind:WAV audio file" and "Bit Rate: 1411 kbps" for FLAC files. I can stream to my web browser.

> [2023-11-01 09:27:09] [DEBUG]       db: Starting query 'SELECT f.* FROM files f JOIN playlistitems pi ON f.path = pi.filepath WHERE f.disabled = 0 AND f.data_kind = 0 AND pi.playlistid = 12 ORDER BY pi.id ASC ;'
> [2023-11-01 09:27:09] [DEBUG]       db: End of query results
> [2023-11-01 09:27:09] [DEBUG]    httpd: Gzipping response
> [2023-11-01 09:27:09] [DEBUG]      rsp: RSP request '/rsp/db/0?query=id%3D12702&type=full' in worker thread 386760
> [2023-11-01 09:27:09] [DEBUG]       db: Running query 'SELECT COUNT(*) FROM files f WHERE f.disabled = 0 AND (f.id = 12702) AND f.data_kind = 0;'
> [2023-11-01 09:27:09] [DEBUG]       db: Starting query 'SELECT f.* FROM files f WHERE f.disabled = 0 AND (f.id = 12702) AND f.data_kind = 0  ORDER BY f.title_sort ;'
> [2023-11-01 09:27:09] [DEBUG]       db: End of query results
> [2023-11-01 09:27:09] [DEBUG]    httpd: Gzipping response
> [2023-11-01 09:27:09] [DEBUG]      rsp: RSP request '/rsp/stream/12702' in worker thread 386760
> [2023-11-01 09:27:09] [DEBUG]       db: Running query 'SELECT f.* FROM files f WHERE f.id = 12702;'
> [2023-11-01 09:27:09] [ INFO]    httpd: Preparing to transcode /var/autofs/removable/easystore10TB/Music/Travis Larson Band/Travis Larson Band-The New Exhibit/02-And Then.flac
> [2023-11-01 09:27:09] [DEBUG]    xcode: Selected encoder 'libmp3lame MP3 (MPEG audio layer 3)'
> [2023-11-01 09:27:09] [DEBUG]   ffmpeg: tb:1/44100 samplefmt:s16 samplerate:44100 chlayout:stereo
> [2023-11-01 09:27:09] [DEBUG]    xcode: Created 'abuffer' filter: 'time_base=1/44100:sample_rate=44100:sample_fmt=s16:channel_layout=stereo'
> [2023-11-01 09:27:09] [DEBUG]    xcode: Created 'aformat' filter: 'sample_fmts=s16p:sample_rates=44100:channel_layouts=stereo'
> [2023-11-01 09:27:09] [DEBUG]    xcode: Created 'abuffersink' filter: ''
> [2023-11-01 09:27:09] [DEBUG]   ffmpeg: auto-inserting filter 'auto_aresample_0' between the filter 'abuffer' and the filter 'aformat'
> [2023-11-01 09:27:09] [DEBUG]   ffmpeg: ch:2 chl:stereo fmt:s16 r:44100Hz -> ch:2 chl:stereo fmt:s16p r:44100Hz
> [2023-11-01 09:27:09] [ INFO]    httpd: Kicking off streaming for /var/autofs/removable/easystore10TB/Music/Travis Larson Band/Travis Larson Band-The New Exhibit/02-And Then.flac
> [2023-11-01 09:27:09] [DEBUG]    httpd: Got 66037 bytes from transcode; streaming file id 12702

**Comment by ejurgensen** [comment]
Ok, that's progress then. It could sound like the Soundbridge gets some metadata that makes it expect wav, so I will check what is sent to it before streaming.

**Comment by ejurgensen** [comment]
I see the RSP playlist response returns info about the media format, so now I made a change so that the response instead tells the Soundbridge that it is 256 kbit mp3. Please try again with the latest commit from the branch included.

**Comment by frankusb** [comment]
It works perfect.

**Comment by ejurgensen** [comment]
Excellent. Does seeking also work?

To change the implementation from experimental to actual, I have to decide which clients should get what formats. In the branch, all clients get 256 kbit mp3, even iTunes (haven't tested if it works, though). I would like to set some good defaults, so that configuration generally isn't needed. I am always weary about adding config options. What defaults do you suggest for RSP?

**Comment by frankusb** [comment]
The Soundbridge does not support seeking, only next and previous tracks and that does work.

For bit rate default, I would suggest 320kbps. That's what I've been using with mp3fs for a while now. I used to frequent the Soundbridge forums and there were never complaints about mp3 bit rates needing to be reduced that I recall. May as well maximize sound quality.

Are there other default settings you are asking about?

To be clear, I am interested in this feature so that I can set wireless Soundbridges to receive 320kbps MP3 and wired Soundbridges to receive 1411kbps WAV. I have been thinking if the default should be 320kbps MP3 to insure that owntone will work with all Soundbridges regardless of type of network connection.

**Comment by Porco-Rosso** [comment]
Agreed 320kbps is a good default. Lots of literature to say that it's the level where its almost impossible to hear any difference from lossless files. It's also the most common mp3 format to download online.

**Comment by ejurgensen** [comment]
While looking at how to do this, I came across [this](https://github.com/owntone/owntone-server/blob/master/src/transcode.c#L69) which seems to imply that the Soundbridge supports ALAC. Do you know if that is the case @frankusb? If so, would it be a better alternative? You would get lossless, but with less compression than the 320 kbps mp3, so requires more wifi bandwidth (I think the compression is about 50%, so half of 1411 kbps I guess).

**Comment by frankusb** [comment]
You tried ALAC [previously](https://github.com/owntone/owntone-server/issues/1182). I also found that the Soundbridge could not handle ALAC over wireless reliably by dropping some ALAC files in owntone.

**Comment by ejurgensen** [comment]
I've been working some more on this, trying to make it more clean/less hardcoded, so that it can be merged. If you have the opportunity, please give the current rspmp3 branch a try again and let me know if it works.

**Comment by frankusb** [comment]
It compiles and the Soundbridge plays fine. I'll leave it running and let you know if I notice anything.

**Comment by frankusb** [comment]
I did notice that the file size shows the FLAC size rather than the MP3 effective size in the Soundbridge song info. If I recall correctly, it also did this with WAV trans-coding as well. It doesn't affect anything that I can tell.

**Comment by frankusb** [comment]
Ripped CDs stored in FLAC format have zero quality loss transcoding to WAV at 44100 for the record unless there is something else I am unaware of in the transcoding from FLAC to WAV. The slight quality loss is only necessary for the wireless Soundbridges. Wired can have bit perfect playback.

And to be overly pedantic, only the M500, M1000 and M2000 have a codec that can play 44.1kHz. For M1001 and R1000, the audio codec can only do 48kHz. For best quality, one would transcode for those to 48kHz as the 400MHz Blackfin in the Soundbridges have audible artifacts upsampling from 44.1kHz to 48kHz.

If one wanted to be able to provide theoretical best quality, each individual device would need control over transcode output (WAV or MP3) and sample rate (44.1kHz vs 48kHz).

**Comment by ejurgensen** [comment]
> I'm thinking I should make another attempt with ALAC. Seems like it could be a silver bullet.

Scratch that, now I see you said before that "I also found that the Soundbridge could not handle ALAC over wireless reliably by dropping some ALAC files in owntone."

**Comment by frankusb** [comment]
OK, I've got the wired vs wireless worked out. Owntone is currently finding all Soundbridges on the local network and adding them as speakers. The RCP protocal is used to do this. The RCP protocol allows one to query wifi parameters. So, GetConnectedWiFiNetwork or GetWiFiSignalQuality could be used to determine the type of network connection and select FLAC transcoding as WAV or MP3 automatically.

**Comment by frankusb** [comment]
Actually, probably a better option is GetIfConfig. Below shows the return value of a wired and wireless Soundbridge.

```
GetIfConfig
GetIfConfig: B200   Interface: 0 (kWireless) has a dhcp assigned address
GetIfConfig:  MAC Address: 00:0D:4B:0A:04:40
GetIfConfig:   IP Address: 192.168.1.44
GetIfConfig:      Netmask: 255.255.255.0
GetIfConfig:      Gateway: 192.168.1.254
GetIfConfig:        DNS 1: 192.168.1.254
GetIfConfig:        DNS 2: 0.0.0.0

GetIfConfig
GetIfConfig: DM90   Interface: 0 (kEthernet) has a dhcp assigned address
GetIfConfig:  MAC Address: 00:0D:4B:04:4F:4B
GetIfConfig:   IP Address: 192.168.1.46
GetIfConfig:      Netmask: 255.255.255.0
GetIfConfig:      Gateway: 192.168.1.254
GetIfConfig:        DNS 1: 192.168.1.254
GetIfConfig:        DNS 2: 0.0.0.0
GetIfConfig: B200   Interface: 1 (kWireless) is disabled
GetIfConfig:  MAC Address: 00:0D:4B:04:CF:4B
```

**Comment by ejurgensen** [comment]
Yes, that sounds like it could work. It would require some internal redesign, however, because right now the RCP implementation is separated from the streaming.

I took a new stab at ALAC and made some progress. Scouring the internet for clues, I found a ffmpeg option that made it work, at least partially. Not with old versions of ffmpeg (not sure where the cutoff is exactly) and not with alll clients. Unfortunately, iTunes is among those - it only plays a few seconds and then aborts.

Could you try with your Roku's? I made a new branch for this, it is `rsp_daap_format1`.

**Comment by frankusb** [comment]
It does not play on the Soundbridge.

```
[2023-12-05 16:16:06] [ INFO]    httpd: Preparing to transcode /var/autofs/removable/easystore10TB/Music/Pink Floyd/Pink Floyd-The Dark Side Of The Moon (Live at Wembley 1974)/10-Eclipse.flac
[2023-12-05 16:16:06] [DEBUG]    xcode: Selected encoder 'ALAC (Apple Lossless Audio Codec)'
[2023-12-05 16:16:06] [DEBUG]   ffmpeg: tb:1/44100 samplefmt:s16 samplerate:44100 chlayout:stereo
[2023-12-05 16:16:06] [DEBUG]    xcode: Created 'abuffer' filter: 'time_base=1/44100:sample_rate=44100:sample_fmt=s16:channel_layout=stereo'
[2023-12-05 16:16:06] [DEBUG]    xcode: Created 'aformat' filter: 'sample_fmts=s16p:sample_rates=44100:channel_layouts=stereo'
[2023-12-05 16:16:06] [DEBUG]    xcode: Created 'abuffersink' filter: ''
[2023-12-05 16:16:06] [DEBUG]   ffmpeg: auto-inserting filter 'auto_aresample_0' between the filter 'abuffer' and the filter 'aformat'
[2023-12-05 16:16:06] [DEBUG]   ffmpeg: ch:2 chl:stereo fmt:s16 r:44100Hz -> ch:2 chl:stereo fmt:s16p r:44100Hz
[2023-12-05 16:16:06] [ INFO]    httpd: Kicking off streaming for /var/autofs/removable/easystore10TB/Music/Pink Floyd/Pink Floyd-The Dark Side Of The Moon (Live at Wembley 1974)/10-Eclipse.flac
[2023-12-05 16:16:06] [DEBUG]    httpd: Got 123255 bytes from transcode; streaming file id 40178
[2023-12-05 16:16:10] [ WARN]      rsp: Connection to '::ffff:192.168.1.40' was closed
```
The Soundbridge shows
Kind: Apple Lossless audio file
Bit Rate: 1411 kbps
Sample Rate: 44.100 kHz
Size: 23.0 MB

From that list, Bit Rate appears incorrect. Size does not match the actual file size of 14M.

```
ls -lh ~/easystore10TB/Music/Pink\ Floyd/Pink\ Floyd-The\ Dark\ Side\ Of\ The\ Moon\ \(Live\ at\ Wembley\ 1974\)/10-Eclipse.flac 
-rw-r--r-- 1 frank frank 14M Mar 23  2023 '/home/frankbanul/easystore10TB/Music/Pink Floyd/Pink Floyd-The Dark Side Of The Moon (Live at Wembley 1974)/10-Eclipse.flac'
```

**Comment by ejurgensen** [comment]
Darn, looks a bit like the iTunes error, where it disconnects after reading a bit.

Calculating the size of ALAC before encoding is impossible, I would think. Some audio can be compressed a lot, other not so much. The bitrate I just set to the same as wav. I think both are probably just used as metadata, so not likely to be the cause of whatever is the cause of the problem.

Btw, in the branch I also changed the wav implementation so that it no longer resamples. Of course that can make your bandwidth problems worse. There is also new option in the config called "prefer_format" (not per device), so there you can switch between alac, wav and mp3.

**Comment by frankusb** [comment]
In my last effort with alac, I found the flags `-movflags +faststart` to be important. Do these apply to your efforts?

**Comment by frankusb** [comment]
Adding `av_dict_set(&options, "movflags", "faststart", 0);` did not change anything for me.

**Comment by frankusb** [comment]
OK, I've been doing some wireshark captures. First difference I notice between an actual alac and a transcoded flac is a missing Content-Length: 28118563 before the file is served. I have not started digging into the alac format yet but the "header" certainly has differences as well.

[Alac.txt](https://github.com/owntone/owntone-server/files/13575605/Alac.txt)
[FlacXcode.txt](https://github.com/owntone/owntone-server/files/13575606/FlacXcode.txt)

**Comment by frankusb** [comment]
Content-Length seems to just be based on if it's a transcoded file or not. FLAC to MP3 transcode is missing it as well but plays fine while actual MP3s provide Content-Length. So probably nothing there.

**Comment by ejurgensen** [comment]
Good that you found how to set those options, it would be great if you could experiment with them. I didn’t try faststart myself, but the internet told me that “movflags” = “empty_moov” was the thing to do. That did in fact make the encoder (the mp4a muxer) stop complaining, but neither VLC nor ffplay could play the stream. Then I tried with “frag_size” instead, because it seems aimed at streaming, and that was better – no encoder errors and works with VLC/ffplay, but not with iTunes. I didn’t try if using both would fix iTunes.

**Comment by frankusb** [comment]
I tried various combinations. I observed that in some cases, the Soundbridge took longer before it failed. I could also cause the transcode to fail. I would guess the Soundbridge does not support fragmented ALAC.

For the record, I tried faststart, frag_every_frame, empty_moov, separate_moof.

**Comment by ejurgensen** [comment]
> I would guess the Soundbridge does not support fragmented ALAC

Yes, I could also fear this, but I will give it a few more shots with iTunes, maybe I will get lucky.

**Comment by frankusb** [comment]
I may have misused av_dict_set. I called it multiple times and I think that only the last option will be set for a given key.

**Comment by frankusb** [comment]
I don't see a good path forward but if you have any ideas I will try them.

av_dict_set_int(&options, "frag_size", 64*1024*1024, 0); fails in a few seconds
av_dict_set(&options, "movflags", "faststart", 0); results in xcode: Error writing header to output buffer: Invalid argument
av_dict_set(&options, "movflags", "empty_moov", 0); fails in a few seconds
av_dict_set(&options, "movflags", "frag_every_frame", 0); fails immediately
av_dict_set(&options, "movflags", "separate_moof", 0); fails immediately
av_dict_set(&options, "movflags", "empty_moov+frag_every_frame", 0); fails immediately
av_dict_set(&options, "movflags", "empty_moov+faststart", 0); fails in a few seconds
av_dict_set(&options, "movflags", "empty_moov+separate_moof", 0); fails in a few seconds
av_dict_set(&options, "movflags", "separate_moof+frag_every_frame", 0); fails immediately
av_dict_set(&options, "movflags", "separate_moof+faststart", 0); results in xcode: Error writing header to output buffer: Invalid argument
av_dict_set(&options, "movflags", "empty_moov+separate_moof+faststart", 0); fails in a few seconds

**Comment by ejurgensen** [comment]
I have one more idea which is about dropping fragmentation and writing best effort mp4 header and footer myself.

But I realized that more fundamentally whether alac/m4a will be viable depends on how the client deals with such a file:

1. Downloads the first part, then starts playback
2. Downloads start and end, then starts playback
3. Downloads the entire file, then starts playback

The last one would make it not viable, as it would of course be super slow. 

I'm going to test this with iTunes and a regular alac-encoded m4a, so without transcoding. Please do the same with Soundbridge if you can.

**Comment by frankusb** [comment]
I am unsure exactly how I would determine the portions of the file the Soundbridge accesses during playback. But I can rule out 3, the Soundbridge streams the entire time a file is being played. The Soundbridge has 13343k of memory as reported by memstat.

I would assert that the Soundbridge can do 1 or 2 depending on the file format and the change in behavior I observed when encoding with faststart.

**Comment by ejurgensen** [comment]
Yes, the Soundbridge probably just reads the file from the beginning. That's what iTunes does. I found that iTunes will only play mp4's made with ffmpeg if the "faststart" flag is used. This flag means that ffmpeg will move the so-called "moov" atom to the beginning of the file. The "moov" atom is about 7k bytes and has various metadata. Importantly for this, it contains an index (chunk table), which, as I understand, tells the player where to find frames in the file. This is useful for seeking (find frame X at byte position Y). Such an index can of course only be made after encoding the entire file. So that's a problem.

I would think that for linear playback without seeking the index shouldn't be needed, so I made some tests to see if iTunes (and ffmpeg) would still play the file if the index was missing or had dummy values. Unfortunately, that wasn't the case. I'm going to try a few more things. If they don't work out, I guess the only way of achieving alac would be to let the filescanner encode files termporarily, just to save this "moov" atom. Would be a rather heavy operation. :-(

**Comment by frankusb** [comment]
I would pay 7k per file to get lossless Soundbridge playback. When I tried my wireless ALAC playback, the wireless Soundbridge and dedicated wireless access point were one room apart. In trying to get WAV to work, I installed a wired drop on the wall in the same room as the wireless Soundbridge. So now the wireless and Soundbridge and dedicated wireless access point are about 6 foot from each other with line of site. And WAV still doesn't work all the time, just mostly. ALAC drops from 1.4Mbps to around 800kbps. I have doubts that it would be 100% like 320kbps MP3 but who knows.

**Comment by ejurgensen** [comment]
>  would pay 7k per file to get lossless Soundbridge playback

Yes, what I meant by heavy wasn't the 7k, it's that owntone would need to temporarily alac encode all files added to the library. And the initial scan of library with a few thousand tracks would take pretty long, during which the server wouldn't be particularly responsive since it would eat cpu (unless consumption could be limited).

**Comment by frankusb** [comment]
> Yes, what I meant by heavy wasn't the 7k, it's that owntone would need to temporarily alac encode all files added to the library. And the initial scan of library with a few thousand tracks would take pretty long, during which the server wouldn't be particularly responsive since it would eat cpu (unless consumption could be limited).

Again, not something that would bother me as it only does it once per file. But I obviously don't speak for everyone, especially those on non intel processors.

**Comment by ejurgensen** [comment]
Yes, if I go that direction, I would go back to wav as default transcode format, and only if the user changes it to alac do the "scan-encoding".

**Comment by ejurgensen** [comment]
I haven't been able to generate a dummy header that iTunes would accept, so now I've instead experimented with using prepared headers, and that is going better. As a step 1, OwnTone is just looking for the header on the filesystem, so if the file is called `something.flac` it will look for `something.flac.mp4h`. The header for a given file can be generated with this rudimentary script:

```
#!/bin/bash
set -e
MP4NAME="/tmp/alac.m4a"
HEADER="$1.mp4h"
ffmpeg -i "$1" -acodec alac -vn -movflags faststart "$MP4NAME"
POS=`grep --byte-offset --only-matching --text mdat "$MP4NAME" | cut -d: -f1`
let LEN=$POS+4
head -c $LEN "$MP4NAME" > "$HEADER"
rm "$MP4NAME"
```
Could you give this is a go with Soundbridge and a test file? Use the latest commit from the `rsp_daap_format1` branch.

Note that this won't work if OwnTone's `decode_audio_filters` config option is set. So if you use that, disable it first.

**Comment by frankusb** [comment]
The Soundbridge is playing ALAC. The script ran very quickly.

**Comment by frankusb** [comment]
On-the-fly creation of mp4 header v2 does not play on the Soundbridge.

**Comment by ejurgensen** [comment]
Bummer. Thanks for a quick test, though. I suppose the transcoded file is somehow malformed. Could you send both a source file and the transcoded file to me at espenjurgensen@gmail.com? You can easily get the transcoded file with something like:

`curl --header "Accept-Codecs: alac" http://localhost:3689/rsp/stream/44674 -o transcoded.m4a`

Also the ffmpeg version.

**Comment by frankusb** [comment]
Emailed.

**Comment by ejurgensen** [comment]
Got them. I think I found a possible culprit. Please try again with the latest commit (the one about using ipod format) from the branch. If that still doesn't work, then please try downloading the m4a with the curl command, and then run this:

`ffmpeg -i transcoded.m4a -vn -c:a copy -movflags faststart transcoded2.m4a`

What this does is rewrite the mp4 headers. Then add the file it to the library and see if the Soundbridge wants to play it.

**Comment by frankusb** [comment]
The Soundbridge is playing ALAC from FLAC source files. It takes about 5-6 seconds between songs though, not the 2 seconds previously.

**Comment by frankusb** [comment]
The 5-6 second delay is only for the first song. Once the Soundbridge starts playing it must have enough buffer from the previous track and request early enough to mask the transcode time. Of course the only songs I've tried have been less than 6 minutes.

**Comment by frankusb** [comment]
I've been using ALAC and MP3 transcoding for some time now. What I found is that over wireless, ALAC has no buffer underruns with my Soundbridge Radio and access point 6 feet away. But with ALAC, the wireless often completely drops and it skips multiple songs for 15-20 seconds then it recovers and plays. The other failure mode is that after playing for a bit, the wireless fails and does not recover and it will skip songs until I reboot it. Sometimes it plays ALAC wirelessly perfectly for 15-30 minutes.

With wired, ALAC has worked flawlessly except for the first transcode startup time (which isn't that much different from the time it takes the hard drive to spin up if it was sleeping anyway). After that there is no break between songs and I've not noticed any unexpected behavior.

My current use case is that I have a cron job that changes the owntone configuration to MP3 transcoding and restarts owntone before my Soundbridge Radio alarm goes off and then another to change the configuration to ALAC and restarts owntone during the day for use with wired Soundbridges. It would be nice to eliminate the restart If the transcode type could be accessed with JSON.

All in all, pretty nice operation. Lossless audio for the wired devices and flawless MP3 operation on the wireless at the cost of two owntone restarts per day. This is much better than running two owntone servers.

**Comment by ejurgensen** [comment]
Thanks for the update, great to hear it is somewhat working, even though I would have hoped wireless ALAC was more stable. I agree more work is required to make it a proper solution. I'm thinking two things:

1. In Settings > "Remotes and Outputs" have a "stream format" setting for each RCP speaker (and maybe also for DAAP clients, if they can be detected). In the future, it might also be useful for Airplay and Chromecast targets.
2. Have the ALAC headers pre-constructed and cached to avoid startup lag.

I've been working on number 2, but I'm not quite satisfied with my current solution. Currently, I'm saving the headers in cache.db, but I think it would be better if they were stored by themselves. I also notice that for some (exotic) files the header construction fails, and then OwnTone should fall back to WAV. Supporting that also requires some changes.

**Comment by frankusb** [comment]
> 1. In Settings > "Remotes and Outputs" have a "stream format" setting for each RCP speaker (and maybe also for DAAP clients, if they can be detected). In the future, it might also be useful for Airplay and Chromecast targets.

That sounds amazing. Very nice way to do it.

**Comment by ejurgensen** [comment]
I've made new branch for testing a more complete solution, so that includes pre-generation of MP4 headers and extension of the JSON api so you can set stream format for each RCP speaker. It would be great if you could give it a test run. The branch is now `rsp_daap_format3`.

When you enable MP4/alac, OwnTone will begin creating headers for every file in your library. The headers will be saved in /var/cache/owntone/xcode.db. If you have a large library this could take quite a while, and the db can grow to considerable size. So I suggest you monitor this. Also note that if the log level is set to debug the logs will also become quite large.

**Comment by frankusb** [comment]
I cheated and already have it running. I have some feedback.

The first major problem is that once "Kicking off header generation" occurs, it seems to preclude all other access by other threads. Owntone becomes unusable via web or RSP. I then disabled all alac in owntone.conf and all RCP speakers. The function cache_xcode_update_cb is still getting called. I commented out event_active(cache_xcode_prepareev, 0, 0).

Generating all MP4 headers on a 32k song library should take around 35 hours at 4 seconds per song. This would be ok if the rest of the server remained accessible.

I currently have wired Soundbridges using WAV and wireless using MP3 and it works great. Very awesome. I had no idea when I originally asked how much would need to be changed. Isn't that typical software development.

A cosmetic issue is that the file size shown by the Soundbridge playing MP3 is off by orders of magnitude, it showed a 4 minute song as 7kB.

**Comment by ejurgensen** [comment]
> The first major problem is that once "Kicking off header generation" occurs, it seems to preclude all other access by other threads.

Yes, I did look at ways to make a thread low priority, and seems the there is Linux scheduler option for batch processes, so I will try if that helps.

> The function cache_xcode_update_cb is still getting called.

Yeah, that must be bug, will see if I can fix that.

> I had no idea when I originally asked how much would need to be changed. Isn't that typical software development.

Indeed, neither did I. The changes are quite extensive, but I think many of them are general improvements to OwnTone that will later also make it easier to add other features.

**Comment by ejurgensen** [comment]
I've added a commit that attempts to fix the problem with the server becoming unavailable. The fix gives the cache thread low priority with the scheduler. I couldn't test it so easily, so not sure if it works as desired. Let me know what you find.

**Comment by frankusb** [comment]
Considering what the thread is performing, would SCHED_IDLE be more appropriate?

**Comment by frankusb** [comment]
Neither SCHED_BATCH or SCHED_IDLE work. RSP cannot connect and the web interface is incomplete, it shows no songs.

**Comment by frankusb** [comment]
Another option would be to cache on access instead of the complete library.

**Comment by ejurgensen** [comment]
Yes, but that would mean a significant lag when starting new tracks, especially on a slow host. That doesn't seem so great.

I think one reason SCHED_BATCH doesn't work is that ffmpeg is spawning it's own threads for encoding. Normally, of course, ffmpeg wants to encode as quickly as possible. I will check if there are ways to limit it.

**Comment by frankusb** [comment]
I'm going to be surprised if this turns out to be priority related. I've got 4 cores. When the problem occurs, one core is 100% utilized and the other cores are very low usage. I would suspect a lock on the cache database not being released between song processing.

**Comment by frankusb** [comment]
That did not go well.
[owntone.log](https://github.com/owntone/owntone-server/files/13922300/owntone.log)

**Comment by ejurgensen** [comment]
Ah yes. Thanks for the log, all those "Kicking off.." weren't supposed to be there. They mean that more and more encoding parallel jobs were being started, which wasn't the idea. I've added another fix to the branch to prevent that, please try again.

**Comment by frankusb** [comment]
It seems to be working now. RSP and http are accessible meanwhile 1 core is pegged at 100%.

**Comment by frankusb** [comment]
Header generation is complete for a library of 31,366 tracks, mostly FLAC (~90%?).

```
[2024-01-12 15:24:56] [  LOG]    cache: Kicking off header generation
[2024-01-13 02:01:57] [  LOG]    cache: Kicking off header generation
[2024-01-13 02:32:31] [  LOG]    cache: Kicking off header generation
[2024-01-13 07:33:01] [  LOG]    cache: Header generation completed
[2024-01-13 08:21:23] [  LOG]    cache: Kicking off header generation
[2024-01-13 08:21:23] [  LOG]    cache: Header generation completed
[2024-01-13 08:22:33] [  LOG]    cache: Kicking off header generation
[2024-01-13 08:22:33] [  LOG]    cache: Header generation completed
```

**Comment by ejurgensen** [comment]
So about 17 hours? Not bad I guess

**Comment by frankusb** [comment]
> So about 16 hours? Not bad I guess

That's with a Intel Core i5 7200U.

**Comment by ejurgensen** [comment]
How large is xcode.db? And can you see how long it takes for owntone to fetch a header?

**Comment by frankusb** [comment]
xcode.db came out to 409M.

The time to fetch seems to vary. In most cases, it seems like about a second. But then 1 out of 5 or 6 it seems to take what it did before caching, 4-5 seconds.

I'm still getting kick offs for no apparent reason.

```
[2024-01-13 10:36:59] [  LOG]    cache: Kicking off header generation
[2024-01-13 10:37:00] [  LOG]    cache: Header generation completed
[2024-01-13 10:37:11] [  LOG]    cache: Kicking off header generation
[2024-01-13 10:37:12] [  LOG]    cache: Header generation completed
[2024-01-13 10:37:28] [  LOG]    cache: Kicking off header generation
[2024-01-13 10:37:28] [  LOG]    cache: Header generation completed
[2024-01-13 10:38:23] [  LOG]    cache: Kicking off header generation
[2024-01-13 10:38:23] [  LOG]    cache: Header generation completed
[2024-01-13 11:26:42] [  LOG]    cache: Kicking off header generation
[2024-01-13 11:26:42] [  LOG]    cache: Header generation completed
```

**Comment by ejurgensen** [comment]
> But then 1 out of 5 or 6 it seems to take what it did before caching, 4-5 seconds.

Would be good if you could get debug level logging of this, so I can see what is going on

> I'm still getting kick offs for no apparent reason.

It could be a library event or a speaker event. Debug level might also reveal that.

**Comment by frankusb** [comment]
> It could be a library event or a speaker event. Debug level might also reveal that.

A speaker event does seem to be related.

curl -X PUT "http://oberon.local:3689/api/outputs/1654141344" --data "{\"format\": \"wav\"}"

But it seems that (spk->format == MEDIA_FORMAT_ALAC && strcmp(spk->output_type, "RCP/SoundBridge") == 0) is not satisfied. cache_xcode_toggle should not be called. prefer_format is set to wav in the config file. All outputs attached.

```
[2024-01-14 10:37:27] [DEBUG]      web: JSON API request '/api/outputs/1654141344' in worker thread 465337
[2024-01-14 10:37:27] [DEBUG]   player: Status update - status: 2, events: 12, caller: speaker_generic_bh
[2024-01-14 10:37:27] [DEBUG]      web: notify callback reason: 71
[2024-01-14 10:37:27] [DEBUG]      web: JSON API request '/api/outputs/2638591226' in worker thread 465334
[2024-01-14 10:37:27] [DEBUG]   player: Status update - status: 2, events: 12, caller: speaker_generic_bh
[2024-01-14 10:37:27] [DEBUG]      web: notify callback reason: 71
[2024-01-14 10:37:27] [DEBUG]      web: JSON API request '/api/outputs/904748892' in worker thread 465337
[2024-01-14 10:37:27] [DEBUG]   player: Status update - status: 2, events: 12, caller: speaker_generic_bh
[2024-01-14 10:37:27] [DEBUG]      web: notify callback reason: 71
[2024-01-14 10:37:32] [DEBUG]       db: Running query 'SELECT COUNT(*) FROM files f WHERE f.disabled = 0 AND f.data_kind = 0;'
[2024-01-14 10:37:32] [DEBUG]       db: Starting query 'SELECT f.* FROM files f WHERE f.disabled = 0 AND f.data_kind = 0  ORDER BY f.id ;'
[2024-01-14 10:37:33] [DEBUG]       db: End of query results
[2024-01-14 10:37:33] [  LOG]    cache: Kicking off header generation
[2024-01-14 10:37:33] [  LOG]    cache: Header generation completed
```
[outputs.txt](https://github.com/owntone/owntone-server/files/13932269/outputs.txt)

**Comment by frankusb** [comment]
> Would be good if you could get debug level logging of this, so I can see what is going on

I've captured 2 slow starts and 1 fast start. It's surprising that in all cases, there is a false start.

[fast.txt](https://github.com/owntone/owntone-server/files/13932440/fast.txt)
[slow.txt](https://github.com/owntone/owntone-server/files/13932441/slow.txt)

**Comment by frankusb** [comment]
Ah, it's missing the cache hit in the slow case for some reason. Still odd why when it consistently false starts.

**Comment by frankusb** [comment]
I touched the file to get it to generate a header and that does not get it to hit the cache on file play. I am wondering if my xcode database is corrupt due to the previous threading experimentation.
[touch.txt](https://github.com/owntone/owntone-server/files/13932482/touch.txt)

**Comment by ejurgensen** [comment]
Thanks for testing, I've added a commit that I think fix of the issues you observed, e.g. the unnecessary "cache: Kicking off header generation" and the slow starts despite headers existing (I don't think there is an issue with your db). Hope you can take it for another spin.

**Comment by frankusb** [comment]
Yes, everything seems to be resolved except for the false start. Is this expected?

```
[2024-01-14 15:25:12] [DEBUG]      rsp: RSP request '/rsp/stream/32242' in worker thread 475008
[2024-01-14 15:25:12] [DEBUG]       db: Running query 'SELECT f.* FROM files f WHERE f.id = 32242;'
[2024-01-14 15:25:12] [ INFO]    httpd: Preparing to transcode /var/autofs/removable/easystore10TB/Music/George Lynch/George Lynch-Guitars at the End of the World/07-The Wolf.flac
[2024-01-14 15:25:12] [DEBUG]    httpd: Checking if client '192.168.1.40' is a speaker
[2024-01-14 15:25:12] [DEBUG]    cache: Cache header hit (11714 bytes)
[2024-01-14 15:25:12] [DEBUG]    xcode: Selected encoder 'ALAC (Apple Lossless Audio Codec)'
[2024-01-14 15:25:12] [DEBUG]   ffmpeg: tb:1/44100 samplefmt:s16 samplerate:44100 chlayout:stereo
[2024-01-14 15:25:12] [DEBUG]    xcode: Created 'abuffer' filter: 'time_base=1/44100:sample_rate=44100:sample_fmt=s16:channel_layout=stereo'
[2024-01-14 15:25:12] [DEBUG]    xcode: Created 'aformat' filter: 'sample_fmts=s16p:sample_rates=44100:channel_layouts=stereo'
[2024-01-14 15:25:12] [DEBUG]    xcode: Created 'abuffersink' filter: ''
[2024-01-14 15:25:12] [DEBUG]   ffmpeg: auto-inserting filter 'auto_aresample_0' between the filter 'abuffer' and the filter 'aformat'
[2024-01-14 15:25:12] [DEBUG]   ffmpeg: ch:2 chl:stereo fmt:s16 r:44100Hz -> ch:2 chl:stereo fmt:s16p r:44100Hz
[2024-01-14 15:25:12] [ INFO]    httpd: Kicking off streaming for /var/autofs/removable/easystore10TB/Music/George Lynch/George Lynch-Guitars at the End of the World/07-The Wolf.flac
[2024-01-14 15:25:12] [DEBUG]    httpd: Got 74993 bytes from transcode; streaming file id 32242
[2024-01-14 15:25:12] [ WARN]      rsp: Connection to '192.168.1.40' was closed
[2024-01-14 15:25:12] [DEBUG]   ffmpeg: Statistics: 242920 bytes read, 0 seeks
[2024-01-14 15:25:13] [DEBUG]      rsp: RSP request '/rsp/stream/32242' in worker thread 475007
[2024-01-14 15:25:13] [DEBUG]    httpd: Found Range header: bytes=144-
[2024-01-14 15:25:13] [DEBUG]       db: Running query 'SELECT f.* FROM files f WHERE f.id = 32242;'
[2024-01-14 15:25:13] [ INFO]    httpd: Preparing to transcode /var/autofs/removable/easystore10TB/Music/George Lynch/George Lynch-Guitars at the End of the World/07-The Wolf.flac
[2024-01-14 15:25:13] [DEBUG]    httpd: Checking if client '192.168.1.40' is a speaker
[2024-01-14 15:25:13] [DEBUG]    cache: Cache header hit (11714 bytes)
[2024-01-14 15:25:13] [DEBUG]    xcode: Selected encoder 'ALAC (Apple Lossless Audio Codec)'
[2024-01-14 15:25:13] [DEBUG]   ffmpeg: tb:1/44100 samplefmt:s16 samplerate:44100 chlayout:stereo
[2024-01-14 15:25:13] [DEBUG]    xcode: Created 'abuffer' filter: 'time_base=1/44100:sample_rate=44100:sample_fmt=s16:channel_layout=stereo'
[2024-01-14 15:25:13] [DEBUG]    xcode: Created 'aformat' filter: 'sample_fmts=s16p:sample_rates=44100:channel_layouts=stereo'
[2024-01-14 15:25:13] [DEBUG]    xcode: Created 'abuffersink' filter: ''
[2024-01-14 15:25:13] [DEBUG]   ffmpeg: auto-inserting filter 'auto_aresample_0' between the filter 'abuffer' and the filter 'aformat'
[2024-01-14 15:25:13] [DEBUG]   ffmpeg: ch:2 chl:stereo fmt:s16 r:44100Hz -> ch:2 chl:stereo fmt:s16p r:44100Hz
[2024-01-14 15:25:13] [DEBUG]    httpd: Stream request with range 144-0
[2024-01-14 15:25:13] [ INFO]    httpd: Kicking off streaming for /var/autofs/removable/easystore10TB/Music/George Lynch/George Lynch-Guitars at the End of the World/07-The Wolf.flac
[2024-01-14 15:25:13] [DEBUG]    httpd: Got 74993 bytes from transcode; streaming file id 32242
[2024-01-14 15:25:13] [DEBUG]    httpd: Got 74368 bytes from transcode; streaming file id 32242
[2024-01-14 15:25:13] [DEBUG]    httpd: Got 67968 bytes from transcode; streaming file id 32242
[2024-01-14 15:25:14] [DEBUG]    httpd: Got 68980 bytes from transcode; streaming file id 32242
[2024-01-14 15:25:14] [DEBUG]    httpd: Got 67951 bytes from transcode; streaming file id 32242
[2024-01-14 15:25:15] [DEBUG]    httpd: Got 70750 bytes from transcode; streaming file id 32242
[2024-01-14 15:25:15] [DEBUG]    httpd: Got 65915 bytes from transcode; streaming file id 32242
[2024-01-14 15:25:15] [DEBUG]    httpd: Got 73847 bytes from transcode; streaming file id 32242
```

**Comment by ejurgensen** [comment]
I'm not sure what to make of that. As you can see, the Soundbridge seems to be streaming a bit, then hanging up and then reconnecting and asking for "Range header: bytes=144-". Can you try with an actual ALAC file for comparison? I wonder if it does the same.

**Comment by frankusb** [comment]
Good test, it does similar.

```
[2024-01-14 16:47:06] [DEBUG]      rsp: RSP request '/rsp/stream/10085' in worker thread 478324
[2024-01-14 16:47:06] [DEBUG]       db: Running query 'SELECT f.* FROM files f WHERE f.id = 10085;'
[2024-01-14 16:47:06] [ INFO]    httpd: Preparing to stream /var/autofs/removable/easystore10TB/Music/2wo/01-I Am a Pig.m4a
[2024-01-14 16:47:06] [ INFO]    httpd: Kicking off streaming for /var/autofs/removable/easystore10TB/Music/2wo/01-I Am a Pig.m4a
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: (LOGGING SKIPPED - above log message is repeating)
[2024-01-14 16:47:06] [ WARN]      rsp: Connection to '192.168.1.40' was closed
[2024-01-14 16:47:06] [DEBUG]      rsp: RSP request '/rsp/stream/10085' in worker thread 478324
[2024-01-14 16:47:06] [DEBUG]    httpd: Found Range header: bytes=1235992-
[2024-01-14 16:47:06] [DEBUG]       db: Running query 'SELECT f.* FROM files f WHERE f.id = 10085;'
[2024-01-14 16:47:06] [ INFO]    httpd: Preparing to stream /var/autofs/removable/easystore10TB/Music/2wo/01-I Am a Pig.m4a
[2024-01-14 16:47:06] [DEBUG]    httpd: Stream request with range 1235992-0
[2024-01-14 16:47:06] [ INFO]    httpd: Kicking off streaming for /var/autofs/removable/easystore10TB/Music/2wo/01-I Am a Pig.m4a
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: Read 4096 bytes; streaming file id 10085
[2024-01-14 16:47:06] [DEBUG]    httpd: (LOGGING SKIPPED - above log message is repeating)
[2024-01-14 16:47:10] [ WARN]      rsp: Connection to '192.168.1.40' was closed
```
Bad news is there appears to be another problem. When I removed a FLAC file to replace it with an ALAC, it seemed to drop all songs.

```
[2024-01-14 15:31:10] [  LOG]      lib: Library init scan completed in 17 sec (0 changes)
[2024-01-14 16:16:47] [  LOG]    cache: Deleting xcode file id 10085
[2024-01-14 16:16:48] [  LOG]    cache: Deleting xcode file id 10086
...
[2024-01-14 16:32:19] [  LOG]    cache: Deleting xcode file id 38130
[2024-01-14 16:32:19] [  LOG]    cache: Deleting xcode file id 38131
[2024-01-14 16:32:19] [  LOG]    cache: Deleting xcode file id 38132
[2024-01-14 16:32:19] [  LOG]    cache: Kicking off header generation
[2024-01-14 16:32:19] [  LOG]    cache: Header generation completed
```

After that the cache was no longer used.

A subsequent restart then added them all back and started preparing headers.

```
[2024-01-14 16:36:03] [DEBUG]    cache: Preparing mp4 header for '/var/autofs/removable/easystore10TB/Music/Roger Hodgson/Roger Hodgson-Classics Live/06-The Logical Song.flac' (file id 10088)
[2024-01-14 16:36:05] [DEBUG]    cache: Preparing mp4 header for '/var/autofs/removable/easystore10TB/Music/Roger Hodgson/Roger Hodgson-Classics Live/02-Give a Little Bit.flac' (file id 10089)
```

**Comment by ejurgensen** [comment]
> When I removed a FLAC file to replace it with an ALAC, it seemed to drop all songs.

Just so I reproduce in the right way: How exactly did you remove/replace? And the server was running, right?

**Comment by frankusb** [comment]
> Just so I reproduce in the right way: How exactly did you remove/replace? And the server was running, right?

I copied the FLAC to another directory not monitored by owntone, ran ffmpeg to create an ALAC, copied the ALAC to the directory monitored by owntone, then deleted the FLAC. Owntone was running the entire time.

I was unable to reproduce myself but thought I better report it.

**Comment by ejurgensen** [comment]
I could reproduce the problem with the headers being dropped, and I believe it should now be fixed.

**Comment by frankusb** [comment]
I've got it compiled and will switch to ALAC for the wired Soundbridges.

**Comment by frankusb** [comment]
ALAC headers finished this morning. I'm getting a lot of Soundbridge reboots. It is reproducible, the same song will always cause the Soundbridge to reboot if I try to play it.

```
[2024-01-20 11:57:30] [DEBUG]       db: Starting query 'SELECT f.* FROM files f WHERE f.disabled = 0 AND (f.album_artist = 'Def Leppard' AND f.album = 'Drastic Symphonies') AND f.data_kind = 0  ORDER BY f.title_sort ;'
[2024-01-20 11:57:30] [DEBUG]       db: End of query results
[2024-01-20 11:57:30] [DEBUG]    httpd: Gzipping response
[2024-01-20 11:57:34] [DEBUG]      rsp: RSP request '/rsp/db/0?query=id%3D22457&type=full' in worker thread 757456
[2024-01-20 11:57:34] [DEBUG]    httpd: Checking if client '192.168.1.40' is a speaker
[2024-01-20 11:57:34] [DEBUG]       db: Running query 'SELECT COUNT(*) FROM files f WHERE f.disabled = 0 AND (f.id = 22457) AND f.data_kind = 0;'
[2024-01-20 11:57:34] [DEBUG]       db: Starting query 'SELECT f.* FROM files f WHERE f.disabled = 0 AND (f.id = 22457) AND f.data_kind = 0  ORDER BY f.title_sort ;'
[2024-01-20 11:57:34] [DEBUG]       db: End of query results
[2024-01-20 11:57:34] [DEBUG]    httpd: Gzipping response
[2024-01-20 11:57:34] [DEBUG]      rsp: RSP request '/rsp/stream/22457' in worker thread 757456
[2024-01-20 11:57:34] [DEBUG]       db: Running query 'SELECT f.* FROM files f WHERE f.id = 22457;'
[2024-01-20 11:57:34] [ INFO]    httpd: Preparing to transcode /var/autofs/removable/easystore10TB/Music/Def Leppard/Def Leppard-Drastic Symphonies/04-Pour Some Sugar On Me (Stripped version).flac
[2024-01-20 11:57:34] [DEBUG]    httpd: Checking if client '192.168.1.40' is a speaker
[2024-01-20 11:57:34] [DEBUG]    cache: Cache header hit (9002 bytes)
[2024-01-20 11:57:34] [DEBUG]    xcode: Selected encoder 'ALAC (Apple Lossless Audio Codec)'
[2024-01-20 11:57:34] [DEBUG]   ffmpeg: tb:1/44100 samplefmt:s16 samplerate:44100 chlayout:stereo
[2024-01-20 11:57:34] [DEBUG]    xcode: Created 'abuffer' filter: 'time_base=1/44100:sample_rate=44100:sample_fmt=s16:channel_layout=stereo'
[2024-01-20 11:57:34] [DEBUG]    xcode: Created 'aformat' filter: 'sample_fmts=s16p:sample_rates=44100:channel_layouts=stereo'
[2024-01-20 11:57:34] [DEBUG]    xcode: Created 'abuffersink' filter: ''
[2024-01-20 11:57:34] [DEBUG]   ffmpeg: auto-inserting filter 'auto_aresample_0' between the filter 'abuffer' and the filter 'aformat'
[2024-01-20 11:57:34] [DEBUG]   ffmpeg: ch:2 chl:stereo fmt:s16 r:44100Hz -> ch:2 chl:stereo fmt:s16p r:44100Hz
[2024-01-20 11:57:34] [ INFO]    httpd: Kicking off streaming for /var/autofs/removable/easystore10TB/Music/Def Leppard/Def Leppard-Drastic Symphonies/04-Pour Some Sugar On Me (Stripped version).flac
[2024-01-20 11:57:34] [DEBUG]    httpd: Got 69253 bytes from transcode; streaming file id 22457
[2024-01-20 11:57:34] [ WARN]      rsp: Connection to '192.168.1.40' was closed
[2024-01-20 11:57:34] [DEBUG]   ffmpeg: Statistics: 280704 bytes read, 0 seeks
[2024-01-20 11:57:34] [DEBUG]      rsp: RSP request '/rsp/stream/22457' in worker thread 757453
[2024-01-20 11:57:34] [DEBUG]    httpd: Found Range header: bytes=144-
[2024-01-20 11:57:34] [DEBUG]       db: Running query 'SELECT f.* FROM files f WHERE f.id = 22457;'
[2024-01-20 11:57:34] [ INFO]    httpd: Preparing to transcode /var/autofs/removable/easystore10TB/Music/Def Leppard/Def Leppard-Drastic Symphonies/04-Pour Some Sugar On Me (Stripped version).flac
[2024-01-20 11:57:34] [DEBUG]    httpd: Checking if client '192.168.1.40' is a speaker
[2024-01-20 11:57:34] [DEBUG]    cache: Cache header hit (9002 bytes)
[2024-01-20 11:57:34] [DEBUG]    xcode: Selected encoder 'ALAC (Apple Lossless Audio Codec)'
[2024-01-20 11:57:34] [DEBUG]   ffmpeg: tb:1/44100 samplefmt:s16 samplerate:44100 chlayout:stereo
[2024-01-20 11:57:34] [DEBUG]    xcode: Created 'abuffer' filter: 'time_base=1/44100:sample_rate=44100:sample_fmt=s16:channel_layout=stereo'
[2024-01-20 11:57:34] [DEBUG]    xcode: Created 'aformat' filter: 'sample_fmts=s16p:sample_rates=44100:channel_layouts=stereo'
[2024-01-20 11:57:34] [DEBUG]    xcode: Created 'abuffersink' filter: ''
[2024-01-20 11:57:34] [DEBUG]   ffmpeg: auto-inserting filter 'auto_aresample_0' between the filter 'abuffer' and the filter 'aformat'
[2024-01-20 11:57:34] [DEBUG]   ffmpeg: ch:2 chl:stereo fmt:s16 r:44100Hz -> ch:2 chl:stereo fmt:s16p r:44100Hz
[2024-01-20 11:57:34] [DEBUG]    httpd: Stream request with range 144-0
[2024-01-20 11:57:34] [ INFO]    httpd: Kicking off streaming for /var/autofs/removable/easystore10TB/Music/Def Leppard/Def Leppard-Drastic Symphonies/04-Pour Some Sugar On Me (Stripped version).flac
[2024-01-20 11:57:34] [DEBUG]    httpd: Got 69253 bytes from transcode; streaming file id 22457
[2024-01-20 11:57:35] [DEBUG]    httpd: Got 73998 bytes from transcode; streaming file id 22457
[2024-01-20 11:57:35] [DEBUG]    httpd: Got 69234 bytes from transcode; streaming file id 22457
[2024-01-20 11:57:49] [ WARN]      rsp: Connection to '192.168.1.40' was closed
[2024-01-20 11:57:49] [DEBUG]   ffmpeg: Statistics: 411776 bytes read, 0 seeks
[2024-01-20 11:57:50] [DEBUG]      rsp: RSP request '/rsp/info' in worker thread 757456
[2024-01-20 11:57:50] [DEBUG]       db: Running query 'SELECT COUNT(*) FROM files f WHERE f.disabled = 0 AND f.data_kind = 0;'
```

**Comment by frankusb** [comment]
I did some more tests. I would assert my headers are corrupt for some files.

`curl --header "Accept-Codecs: alac" http://localhost:3689/rsp/stream/22457 -o transcoded.m4a`

resulted in an unplayable file in VLC. MediaInfo showed it was truncated.

`touch ../easystore10TB/Music/Def\ Leppard/Def\ Leppard-Drastic\ Symphonies/04-Pour\ Some\ Sugar\ On\ Me\ \(Stripped\ version\).flac`

```
[2024-01-21 10:51:47] [  LOG]    cache: Deleting xcode file id 22457
[2024-01-21 10:51:47] [  LOG]    cache: Adding xcode file id 22457, path '/var/autofs/removable/easystore10TB/Music/Def Leppard/Def Leppard-Drastic Symphonies/04-Pour Some Sugar On Me (Stripped version).flac'
[2024-01-21 10:51:47] [  LOG]    cache: Kicking off header generation
[2024-01-21 10:51:50] [  LOG]    cache: Header generation completed
```

`curl --header "Accept-Codecs: alac" http://localhost:3689/rsp/stream/22457 -o transcoded22457.m4a`

resulted in a playable file in VLC and no truncation in MediaInfo

**Comment by hacketiwack** [comment]
> I did some more tests. I would assert my headers are corrupt for some files.

Not completely related to your situation, however, in my experience, I rarely had good result converting to ALAC with ffmpeg. For that, I use a script I created for `afconvert`. And of course it works only under macOS :-D. That's why I ended up converting all my lossless collection to ALAC in m4a containers.

This is the script I use:

```zsh
#!/bin/zsh

# Format flags
flags=([16]=1 [20]=2 [24]=3 [32]=4)

# Retrieve the bit depth of the audio file
depth=$(afinfo "$1" | sed -n 's/.*depth: I\(.*\)/\1/p')

# Convert the audio file
echo "Converting $1"
afconvert -f m4af -d alac/${flags[depth]} "$1" "${1%.*}.m4a"
```

You'll have to retag your files afterwards.

**Comment by ejurgensen** [comment]
@frankusb good idea using "touch" to recreate the header and then checking again. To me that indicates a bug in the cache storage/retrieval of headers. I will see if I can catch it.

**Comment by frankusb** [comment]
@ejurgensen if you are not interested in any of my present database information I was planning on removing it this weekend and rebuilding to see if the changes over time contributed to the current state or if there is something in the current codebase. I could always preserve it if you feel it might be interesting later.

**Comment by ejurgensen** [comment]
I haven't gotten around to looking at the latest issue yet, so not sure how much more information I might need.

If you currently have a file which doesn't play due to invalid header, I think it would be great if you could save that invalid header. I think you can do that like this:

```
sqlite3 /var/cache/owntone/xcode.db
select id, file_id, format, length(header). hex(header) from data where file_id = [file_id];
```

Then just save the output. Then, when you have rebuilt, check if the new header matches. Also check if any other headers match:

`select id, file_id, format, length(header). hex(header) from data where length(header) = [header_len];`

**Comment by frankusb** [comment]
I'm happy to leave the xcode database as is, weekends are just a good time to run the rebuild.

**Comment by ejurgensen** [comment]
Hi @frankusb I found a possible issue, so I've made some modifications to the branch that I hope you can test. Before doing so, you would have to delete your current xcode.db and rebuild. The rebuilding should be 4x faster now, since one of the modifications is to run the encoding in parallel. I've tried to do it in a way that hopefully doesn't make the server inoperable during the encoding, but let me know what your experience is. You can adjust the number for parallel jobs with CACHE_XCODE_NTHREADS.

I tried testing the validity of the files with ffprobe, and they all pass, except for some special formats that can't be transcoded (e.g. due to DRM). If your Soundbridge still doesn't like some of the files, please check it with ffprobe, so I can see if that works as a test.

**Comment by frankusb** [comment]
I gave it a go. All good as far as I can tell. I can't find any problem files for the Soundbridge but will keep it set to ALAC.

```
[2024-01-28 08:27:44] [  LOG]    cache: Kicking off header generation
[2024-01-28 09:17:42] [  LOG]    xcode: Error while feeding the filtergraph: Invalid argument
[2024-01-28 09:17:45] [  LOG]    xcode: Error while feeding the filtergraph: Invalid argument
[2024-01-28 10:02:34] [  LOG]   ffmpeg: No JPEG data found in image
...
[2024-01-28 13:25:05] [  LOG]   ffmpeg: (LOGGING SKIPPED - above log message is repeating)
[2024-01-28 15:05:24] [  LOG]    cache: Header generation completed
```

Some humble feedback. It might be useful to include the file for the above errors so one could correct/examine them. I happen to have 4 cores so your CACHE_XCODE_NTHREADS choice matches nicely. Querying the number of cores would probably be better. I could tell the system was sluggish but nothing stopped working. SCHED_IDLE vs SCHED_BATCH may help with this.

**Comment by ejurgensen** [comment]
> It might be useful to include the file for the above errors so one could correct/examine them

Yes, I've added a fix for that now

**Comment by frankusb** [comment]
The fix did not trigger for me.

```
[2024-01-30 18:10:47] [  LOG]    xcode: Error while feeding the filtergraph: Invalid argument
[2024-01-30 18:10:49] [  LOG]    xcode: Error while feeding the filtergraph: Invalid argument
[2024-01-30 18:55:15] [  LOG]   ffmpeg: No JPEG data found in image
```

I've checked to make sure the update is in my codebase, make, install.

**Comment by ejurgensen** [comment]
Seems to work ok when I mock the error:

```
[2024-02-01 14:38:28] [  LOG]    xcode: Error while feeding the filtergraph: Invalid argument
[2024-02-01 14:38:28] [  LOG]    cache: Error preparing mp4 header for 'path/to/file.mp3' (file id 15)
```

**Comment by frankusb** [comment]
filter_encode_write can be called from decode_filter_encode_write and transcode_encode. Is it possible that only one return path results in the additional log message? I can't see why it's not in my log.

On the bright side, I used a Soundbridge to play ALAC for 2-3 hours last night and it worked perfectly.

**Comment by ejurgensen** [comment]
Yes, you are right, there are some code paths that could lead to no logging of the file name. It seems to be when the encoder is flushed, but I'm not sure if that is the only way. I'll add some extra logging so we can check what is leading to it.

**Comment by frankusb** [comment]
I just noticed that WAV playtback is not working on the latest version of the rsp_daap_format3 branch. Sorry I took so long to notice, the problem only occurs at the very end of the first track played. When the Soundbridge buffers the next song during the end of the first track, it reboots.

Once ALAC was working, I switched all of my wired Soundbridges to ALAC and never retested WAV. I should have left half in WAV and half in ALAC. I found this because I added a Soundbridge to my network and it defaulted to WAV. Switching it to ALAC allows it to play multiple tracks without rebooting.

**Comment by ejurgensen** [comment]
Thanks for the notice, it's great that you are testing. I haven't done much with the branch in some time, exactly because the changes are pretty extensive and thus need a lot of testing, otherwise merging into master will be too risky. I've also been busy with other things, but I haven't forgotten about this.

**Comment by frankusb** [comment]
ALAC and MP3 have been flawless on the most recent version of the branch. I've been using it exclusively. I've done lots of library changes, playlist changes and general usage.

**Comment by ejurgensen** [comment]
Good to hear. Also about the general usage, because it is also general regressions I'm concerned about.

**Comment by hacketiwack** [comment]
<off-topic>
Not directly related to the problem here, I'm wondering what is the advantage of streaming PCM audio (aka Wav files) instead of ALAC audio, which is also lossless?
</off-topic>

**Comment by frankusb** [comment]
@hacketiwack ALAC saves network bandwidth is the only advantage from my perspective. This is mostly negligible though.

@ejurgensen my general usage includes Airplay speakers (receivers actually), Apple remote control, Web interface remote control, Soundbridge RSP, and a little Chromecast speakers (they tend to cut off but always have).

**Comment by ejurgensen** [comment]
PCM/WAV has the advantage that it is decode only, so less cpu required, and also OwnTone can decode on the fly. With ALAC, OwnTone has to prebuild a cache of ALAC headers. This is cpu intensive and the cached headers do end up filling some disk space.

**Comment by frankusb** [comment]
> I just noticed that WAV playtback is not working on the latest version of the rsp_daap_format3 branch. Sorry I took so long to notice, the problem only occurs at the very end of the first track played. When the Soundbridge buffers the next song during the end of the first track, it reboots.

I need to amend this report as it manifests on only one M1001 that I have found so far, it does not reproduce on an M2000 in WAV mode. Even more insidious, it appears to be length of track related, the M1001 plays a 0:52 song while in WAV mode but fails on 3:xx and 4:xx songs. I will leave all of my Soundbridges in WAV mode for a while to see if I can narrow this down. It's possible this is device specific?

**Comment by ejurgensen** [comment]
It's a hard to say. A possibility is that there is something in the WAV header that the M1001 doesn't like, but that the M2000 accepts. The header has the length of the track, so that could explain why that makes a difference. Here's how the header is made fyi: https://github.com/owntone/owntone-server/blob/master/src/transcode.c#L439

**Comment by ejurgensen** [comment]
I've merged this in now (merge commit 4cbce79). It's a large change, so it's with some trepidation, but otherwise it would just become stale.

**Comment by frankusb** [comment]
Oh wow, time for me to move to master then. I will do so tonight. The default of WAV and the M1001 reboot issue probably needs to be figured out then. If you could point me to some places to instrument, I can do some testing.

**Comment by ejurgensen** [comment]
Yes, I better look at that. I created a bug report for that as you can see above.

**Comment by frankusb** [comment]
After updating to test [1798](https://github.com/owntone/owntone-server/issues/1798), mp3 transcoding is no longer working. I believe the setting is correct in the outputs.

`{ "id": "2247276689", "name": "SBRadioBedroom", "type": "RCP\/SoundBridge", "selected": false, "has_password": false, "requires_auth": false, "needs_auth_key": false, "volume": 50, "format": "mp3", "supported_formats": [ "wav", "mp3", "alac" ] }`

But the type shows "WAV" on the Soundbridge.

**Comment by frankusb** [comment]
Never mind, I went to try to figure out which committ was causing this and now I am back on origin/master and transcoding to mp3 is working fine. My Soundbridge Radio has been buffering for the past two days so I know it wasn't working then but I can't break it again.
